client/python/gradio_client/client.py

```diff
@@ -0,0 +1,300 @@
+from __future__ import annotations
+
+import concurrent.futures
+import json
+import re
+import threading
+import uuid
+from concurrent.futures import Future
+from typing import Any, Callable, Dict, List, Tuple
+
+import requests
+import websockets
+from gradio_client import serializing, utils
+from gradio_client.serializing import Serializable
+from huggingface_hub.utils import build_hf_headers, send_telemetry
+from packaging import version
+
+
+class Client:
+    def __init__(
+        self,
+        space: str | None = None,
+        src: str | None = None,
+        hf_token: str | None = None,
+        max_workers: int = 40,
+    ):
+        self.hf_token = hf_token
+        self.headers = build_hf_headers(
+            token=hf_token,
+            library_name="gradio_client",
+            library_version=utils.__version__,
+        )
+
+        if space is None and src is None:
+            raise ValueError("Either `space` or `src` must be provided")
+        elif space and src:
+            raise ValueError("Only one of `space` or `src` should be provided")
+        self.src = src or self._space_name_to_src(space)
+        if self.src is None:
+            raise ValueError(
+                f"Could not find Space: {space}. If it is a private Space, please provide an hf_token."
+            )
+        else:
+            print(f"Loaded as API: {self.src} ✔")
+
+        self.api_url = utils.API_URL.format(self.src)
+        self.ws_url = utils.WS_URL.format(self.src).replace("https", "wss")
+        self.config = self._get_config()
+
+        self.endpoints = [
+            Endpoint(self, fn_index, dependency)
+            for fn_index, dependency in enumerate(self.config["dependencies"])
+        ]
+
+        # Create a pool of threads to handle the requests
+        self.executor = concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)
+
+        # Disable telemetry by setting the env variable HF_HUB_DISABLE_TELEMETRY=1
+        threading.Thread(target=self._telemetry_thread).start()
+
+    def predict(
+        self,
+        *args,
+        api_name: str | None = None,
+        fn_index: int = 0,
+        result_callbacks: List[Callable] | None = None,
+    ) -> Future:
+        if api_name:
+            fn_index = self._infer_fn_index(api_name)
+
+        end_to_end_fn = self.endpoints[fn_index].end_to_end_fn
+        future = self.executor.submit(end_to_end_fn, *args)
+        job = Job(future)
+
+        if result_callbacks:
+
+            def create_fn(callback) -> Callable:
+                def fn(future):
+                    if isinstance(future.result(), tuple):
+                        callback(*future.result())
+                    else:
+                        callback(future.result())
+
+                return fn
+
+            for callback in result_callbacks:
+                job.add_done_callback(create_fn(callback))
+
+        return job
+
+    ##################################
```

Comment:
dont like this big comment block, private methods are self evident via underscore prefix

Reply:
Sg